File: templates/csharp/protocols/param_codec.cs
using System.Buffers.Binary;
using Mavlink.Dialects;

namespace Mavlink;

/// <summary>
/// Byte-wise parameter value encoding per MAVLink parameter protocol.
/// </summary>
/// <remarks>See https://mavlink.io/en/services/parameter.html</remarks>
public static class ParamCodec
{
    public static float EncodeInt8(int value) => EncodeInt32(value);

    public static int DecodeInt8(float encoded) => DecodeInt32(encoded);

    public static float EncodeUint8(int value) => EncodeUint32(value);

    public static int DecodeUint8(float encoded) => DecodeUint32(encoded);

    public static float EncodeInt16(int value) => EncodeInt32(value);

    public static int DecodeInt16(float encoded) => DecodeInt32(encoded);

    public static float EncodeUint16(int value) => EncodeUint32(value);

    public static int DecodeUint16(float encoded) => DecodeUint32(encoded);

    public static float EncodeInt32(int value)
    {
        Span<byte> bytes = stackalloc byte[4];
        BinaryPrimitives.WriteInt32LittleEndian(bytes, value);
        return BinaryPrimitives.ReadSingleLittleEndian(bytes);
    }

    public static int DecodeInt32(float encoded)
    {
        Span<byte> bytes = stackalloc byte[4];
        BinaryPrimitives.WriteSingleLittleEndian(bytes, encoded);
        return BinaryPrimitives.ReadInt32LittleEndian(bytes);
    }

    public static float EncodeUint32(int value)
    {
        Span<byte> bytes = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32LittleEndian(bytes, (uint)value);
        return BinaryPrimitives.ReadSingleLittleEndian(bytes);
    }

    public static int DecodeUint32(float encoded)
    {
        Span<byte> bytes = stackalloc byte[4];
        BinaryPrimitives.WriteSingleLittleEndian(bytes, encoded);
        return (int)BinaryPrimitives.ReadUInt32LittleEndian(bytes);
    }

    public static float EncodeFloat(float value) => value;

    public static float DecodeFloat(float encoded) => encoded;

    public static float EncodeValue(decimal value, MavParamType type) =>
        type switch
        {
            MavParamType.MAV_PARAM_TYPE_UINT8 => EncodeUint8((int)value),
            MavParamType.MAV_PARAM_TYPE_INT8 => EncodeInt8((int)value),
            MavParamType.MAV_PARAM_TYPE_UINT16 => EncodeUint16((int)value),
            MavParamType.MAV_PARAM_TYPE_INT16 => EncodeInt16((int)value),
            MavParamType.MAV_PARAM_TYPE_INT32 => EncodeInt32((int)value),
            MavParamType.MAV_PARAM_TYPE_UINT32 => EncodeUint32((int)value),
            MavParamType.MAV_PARAM_TYPE_REAL32 => EncodeFloat((float)value),
            _ => (float)value,
        };

    public static decimal DecodeValue(float encoded, MavParamType type) =>
        type switch
        {
            MavParamType.MAV_PARAM_TYPE_UINT8 => DecodeUint8(encoded),
            MavParamType.MAV_PARAM_TYPE_INT8 => DecodeInt8(encoded),
            MavParamType.MAV_PARAM_TYPE_UINT16 => DecodeUint16(encoded),
            MavParamType.MAV_PARAM_TYPE_INT16 => DecodeInt16(encoded),
            MavParamType.MAV_PARAM_TYPE_INT32 => DecodeInt32(encoded),
            MavParamType.MAV_PARAM_TYPE_UINT32 => DecodeUint32(encoded),
            MavParamType.MAV_PARAM_TYPE_REAL32 => (decimal)DecodeFloat(encoded),
            _ => (decimal)encoded,
        };

    /// <summary>Encode a short ASCII parameter name into a MAVLink <c>char[16]</c> field.</summary>
    public static byte[] ParamIdFromString(string name)
    {
        var id = new byte[16];
        var bytes = System.Text.Encoding.ASCII.GetBytes(name);
        Array.Copy(bytes, 0, id, 0, Math.Min(bytes.Length, 16));
        return id;
    }

    /// <summary>Decode a MAVLink <c>char[16]</c> parameter id to a string.</summary>
    public static string ParamIdToString(byte[] id)
    {
        var end = Array.FindIndex(id, value => value == 0);
        if (end < 0)
        {
            end = id.Length;
        }

        return System.Text.Encoding.ASCII.GetString(id, 0, end);
    }
}

File: templates/ts/protocols/param_codec.ts
import { MavParamType } from '../mavlink';

/** Byte-wise parameter value encoding per MAVLink parameter protocol. */
export class ParamCodec {
  private constructor() {}

  static encodeInt8(value: number): number {
    return ParamCodec.encodeInt32(value);
  }

  static decodeInt8(encoded: number): number {
    return ParamCodec.decodeInt32(encoded);
  }

  static encodeUint8(value: number): number {
    return ParamCodec.encodeUint32(value);
  }

  static decodeUint8(encoded: number): number {
    return ParamCodec.decodeUint32(encoded);
  }

  static encodeInt16(value: number): number {
    return ParamCodec.encodeInt32(value);
  }

  static decodeInt16(encoded: number): number {
    return ParamCodec.decodeInt32(encoded);
  }

  static encodeUint16(value: number): number {
    return ParamCodec.encodeUint32(value);
  }

  static decodeUint16(encoded: number): number {
    return ParamCodec.decodeUint32(encoded);
  }

  static encodeInt32(value: number): number {
    const bytes = new DataView(new ArrayBuffer(4));
    bytes.setInt32(0, value, true);
    return bytes.getFloat32(0, true);
  }

  static decodeInt32(encoded: number): number {
    const bytes = new DataView(new ArrayBuffer(4));
    bytes.setFloat32(0, encoded, true);
    return bytes.getInt32(0, true);
  }

  static encodeUint32(value: number): number {
    const bytes = new DataView(new ArrayBuffer(4));
    bytes.setUint32(0, value >>> 0, true);
    return bytes.getFloat32(0, true);
  }

  static decodeUint32(encoded: number): number {
    const bytes = new DataView(new ArrayBuffer(4));
    bytes.setFloat32(0, encoded, true);
    return bytes.getUint32(0, true);
  }

  static encodeFloat(value: number): number {
    return value;
  }

  static decodeFloat(encoded: number): number {
    return encoded;
  }

  static encodeValue(value: number, type: MavParamType): number {
    switch (type) {
      case MavParamType.MAV_PARAM_TYPE_UINT8:
        return ParamCodec.encodeUint8(value);
      case MavParamType.MAV_PARAM_TYPE_INT8:
        return ParamCodec.encodeInt8(value);
      case MavParamType.MAV_PARAM_TYPE_UINT16:
        return ParamCodec.encodeUint16(value);
      case MavParamType.MAV_PARAM_TYPE_INT16:
        return ParamCodec.encodeInt16(value);
      case MavParamType.MAV_PARAM_TYPE_INT32:
        return ParamCodec.encodeInt32(value);
      case MavParamType.MAV_PARAM_TYPE_UINT32:
        return ParamCodec.encodeUint32(value);
      case MavParamType.MAV_PARAM_TYPE_REAL32:
        return ParamCodec.encodeFloat(value);
      default:
        return value;
    }
  }

  static decodeValue(encoded: number, type: MavParamType): number {
    switch (type) {
      case MavParamType.MAV_PARAM_TYPE_UINT8:
        return ParamCodec.decodeUint8(encoded);
      case MavParamType.MAV_PARAM_TYPE_INT8:
        return ParamCodec.decodeInt8(encoded);
      case MavParamType.MAV_PARAM_TYPE_UINT16:
        return ParamCodec.decodeUint16(encoded);
      case MavParamType.MAV_PARAM_TYPE_INT16:
        return ParamCodec.decodeInt16(encoded);
      case MavParamType.MAV_PARAM_TYPE_INT32:
        return ParamCodec.decodeInt32(encoded);
      case MavParamType.MAV_PARAM_TYPE_UINT32:
        return ParamCodec.decodeUint32(encoded);
      case MavParamType.MAV_PARAM_TYPE_REAL32:
        return ParamCodec.decodeFloat(encoded);
      default:
        return encoded;
    }
  }

  /** Encode a short ASCII parameter name into a MAVLink `char[16]` field. */
  static paramIdFromString(name: string): number[] {
    const id: number[] = [];
    for (const codeUnit of name.slice(0, 16)) {
      id.push(codeUnit.charCodeAt(0));
    }
    while (id.length < 16) {
      id.push(0);
    }
    return id;
  }

  /** Decode a MAVLink `char[16]` parameter id to a string. */
  static paramIdToString(id: number[]): string {
    const end = id.findIndex((value) => value === 0);
    const slice = end === -1 ? id : id.slice(0, end);
    return String.fromCharCode(...slice);
  }
}
